### PR TITLE
implement basic support for type checking of easyconfig parameters

### PR DESCRIPTION
**Pull request by boegel**
Without type checking, you can run into nasty errors, for example the one below when using a non-string value for the `version` easyconfig parameter:

```
Traceback (most recent call last):
  File "/Users/kehoste/work/easybuild-framework/easybuild/main.py", line 361, in <module>
    main()
  File "/Users/kehoste/work/easybuild-framework/easybuild/main.py", line 276, in main
    easyconfigs, generated_ecs = parse_easyconfigs(paths)
  File "/Users/kehoste/work/easybuild-framework/easybuild/framework/easyconfig/tools.py", line 345, in parse_easyconfigs
    ecs = process_easyconfig(ec_file, **kwargs)
  File "/Users/kehoste/work/easybuild-framework/easybuild/framework/easyconfig/easyconfig.py", line 943, in process_easyconfig
    ec = EasyConfig(spec, build_specs=build_specs, validate=validate, hidden=hidden)
  File "/Users/kehoste/work/easybuild-framework/easybuild/framework/easyconfig/easyconfig.py", line 188, in __init__
    self.parse()
  File "/Users/kehoste/work/easybuild-framework/easybuild/framework/easyconfig/easyconfig.py", line 296, in parse
    self.generate_template_values()
  File "/Users/kehoste/work/easybuild-framework/easybuild/framework/easyconfig/easyconfig.py", line 661, in generate_template_values
    self._generate_template_values(skip_lower=True)
  File "/Users/kehoste/work/easybuild-framework/easybuild/framework/easyconfig/easyconfig.py", line 675, in _generate_template_values
    ignore=ignore, skip_lower=skip_lower)
  File "/Users/kehoste/work/easybuild-framework/easybuild/framework/easyconfig/templates.py", line 161, in template_constant_dict
    version = LooseVersion(version).version
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/version.py", line 265, in __init__
    self.parse(vstring)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/distutils/version.py", line 274, in parse
    self.component_re.split(vstring))
TypeError: expected string or buffer
```

With these changes in place, you'll get this instead:

```
EasyBuildError: "Failed to process easyconfig example.eb: Type checking of easyconfig parameter values failed: value for 'version' should be of type 'basestring'"
```

This implementation is intended as a starting point, since only the values for `name`/`version` will be type-checked for now.
Complex values like `sanity_check_paths` will need additional logic/functionality, and support for automatically trying to convert values is probably worthwhile too. For example, with easyconfig files in YAML syntax (see #1407), a value for `version` like `1.6` would be parsed as `float` by default, although it should be a string value...

### CODE REVIEW

**Comment by boegel** [comment]
@wpoely86, @Caylo: please review?

File: easybuild/framework/easyconfig/parser.py [hunk 1]
```diff
@@ -106,6 +107,18 @@ def process(self, filename=None):
         self._read(filename=filename)
         self._set_formatter()
 
+    def check_values_types(self, cfg):
+        """Check types of easyconfig parameter values."""
```
Comment (wpoely86) [inline easybuild/framework/easyconfig/parser.py]: document args

File: easybuild/framework/easyconfig/parser.py [hunk 2]
```diff
@@ -106,6 +107,18 @@ def process(self, filename=None):
...
+        wrong_type_msgs = []
+        for key in cfg:
+            if not check_type_of_param_value(key, cfg[key]):
+                wrong_type_msgs.append("value for '%s' should be of type '%s'" % (key, TYPES[key].__name__))
+
+        if wrong_type_msgs:
+            raise EasyBuildError("Type checking of easyconfig parameter values failed: %s", ', '.join(wrong_type_msgs))
+        else:
+            self.log.info("Type checking of easyconfig parameter values passed!")
```
Comment (wpoely86) [inline easybuild/framework/easyconfig/parser.py]: debug?

**Comment by wpoely86** [comment]
lgtm

File: easybuild/framework/easyconfig/parser.py (hunk 2 above)
Comment (boegel) [inline easybuild/framework/easyconfig/parser.py]: This is just a single log line, it's worth having this at the `info` level to indicate type checking was performed.

**Comment by boegel** [comment]
Thanks for the review @wpoely86 and @Caylo!

**Pull request merged**